Emoncms API now provides a Unit of Measurement

borpin opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
The way the `unit_of_measurement` is set has changed to make use of the ability for this to be set in the source feed in Emocms.

The unit will now be assigned as follows:
 
1. Take the unit from the API if it is available
1. Otherwise take the unit from config, either specified or the default unit (W)

## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
The EmonCMS API has been amended (from V9.9.1) to include a 'unit' as part of it's payload.  By using this information, all the sensors can be created without the need for individual sensors to be setup by type.

If no unit is specified by the Emoncms API, then the specified unit will be use or if there is neither the default of W will be used.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml entry using cloud based Emoncms
sensor:
  platform: emoncms
  api_key: API_KEY
  url: https://emoncms.org
  id: 1
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes 
- This PR is related to issue: #31251 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/12153

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

borpin committed:
Emoncms API provides a Unit

The EmonCMS API has been amended to include a 'unit' as part of it's payload.  By using this information, all the sensors can be created without the need for individual sensors to be setup by type.

The change is backward compatible so if a unit type has been specified in the configuration, this will be used by default.

If no unit is pecified either by the Home Assistant configuration, or the Emoncms API, then the default of W will be used as before.

borpin committed:
Update sensor.py

Check the 'unit' key is in the API call. Older systems may not have that key in the payload.

springstan reviewed (commented):
Please revert some of your changes, because we use the `default` value in voluptuous to validate the config and preserve a default value instead of shifting this logic to if statements.

borpin commented:
If I do it this way, it changes the whole logic and it will be a breaking change which I am trying to avoid.

I need to check and see if a value has been put into the 'units_of_measurement' by the user. I can only do that is the default value has not been loaded at the start (I think).

If value set in 'units_of_measuremant' by user then use it (the current situation)
elsif value exists in API Payload use that
else use the DEFAULT (current situation)

I'd be happy to remove the default out at the start completely.

borpin committed:
Modified approach with new configuration item

borpin commented:
I have modified the approach and used a new config item - I will need to update the documentation PR if this is acceptable.

frenck commented:
Changes to the unit of measurement is a breaking change imho.

borpin commented:
> Changes to the unit of measurement is a breaking change imho.

My modified approach to include a new configuration item, would mean that a user has to make a change to their configuration for it to change the behaviour. The change will therefore, **not** break any existing setups.

borpin commented:
Created a new Documentation PR.

springstan commented on homeassistant/components/emoncms/sensor.py:
```diff
@@ -57,7 +58,8 @@
         vol.Optional(CONF_VALUE_TEMPLATE): cv.template,
-        vol.Optional(CONF_UNIT_OF_MEASUREMENT, default=DEFAULT_UNIT): cv.string,
+        vol.Optional(CONF_UNIT_OF_MEASUREMENT,default=DEFAULT_UNIT): cv.string,
+        vol.Optional(CONF_USE_EMONCMS_UNIT,default=DEFAULT_CONF_USE_EMONCMS_UNIT): cv.boolean,
```
Tbh I would prefer a breaking change instead of introducing such an option. An integration should always use the unit of measurement from the api if it is available. It definitely would be more consistent with other integrations and I cannot think of another one right now, where such an option is used.

borpin replied:
It needs to be your design choice as you have not really liked either of my options.  Because of the DEFAULT_VALUE, it makes it very difficult to migrate cleanly.

Are you saying you want 
1. the option removed
2. The API unit used if present
3. the UNIT_OF_MEASUREMENT that has either been set of where the default has picked it up?

The code would then become...
```
unit_of_measurement = elem.get("unit",config.get(CONF_UNIT_OF_MEASUREMENT))
```

springstan replied:
1. Yes, `CONF_USE_EMONCMS_UNIT` should be removed.
2. Yes, exactly.

> I have missed a use case here, the 'unit' key may exist in the API call but it may not be set in Emoncms so will be just an empty string.

Can the user set the unit in Emoncms manually? Because if that is the case I would even remove the unit of measurement and add a note to the documentation that the user should add the unit in Emoncms manually.

springstan replied:
Any updates on this? 🙂

borpin replied:
Just in. The problem this was added to the API at some point so older instances won't work (and upgrading is not always straight forward so not always an option).

Just trying to check and see when it was added.

For now, I'd rather keep it but mark it as a feature that will go at some point in the future. I've seen that with other integrations - not sure how that message would be added.

Thanks for your patience! 😁

springstan replied:
Okay thanks for looking deepr into this :)

IMHO we should leave your PR as it is, to have the following logic:

1. take the unit from the API if it is available
2. otherwise take the unit from config, either specified or the default one

Consequently, all sensors will have a unit of measurement to make it easier to display them in graphs etc. and the new feature is included.
However, I would suggest to add a breaking change and add a note to the documentation describing the logic and how to upgrade to new firmware or how to set this unit in emoncms.

What do you think about all of this?

borpin replied:
Ok great. I'll amend the PR to show it as a breaking change and update the documentation (again) 😁.

I have also worked out when the feature was added (summer 2018) so I can add that as well.

springstan commented on homeassistant/components/emoncms/manifest.json:
```diff
@@ -4,5 +4,5 @@
   "requirements": [],
   "dependencies": [],
-  "codeowners": []
+  "codeowners": ["@borpin"]
```
Please add yourself to the [codeowners file](https://github.com/home-assistant/home-assistant/blob/dev/CODEOWNERS).

springstan commented on homeassistant/components/emoncms/sensor.py:
```diff
@@ -107,6 +109,9 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
             name = sensor_names.get(int(elem["id"]), None)
 
+        if config.get(CONF_USE_EMONCMS_UNIT):
+            unit_of_measurement = elem.get("unit","")
```
Furthermore, if no unit of measurement is found here an empty string is used which is not ideal.
Either set it to `None` or use the unit that has been used before for all sensors.

borpin replied:
Would using the DEFAULT_UNIT be OK?

springstan replied:
Yes 👍

borpin replied:
As above how about
```
unit_of_measurement = elem.get("unit",config.get(CONF_UNIT_OF_MEASUREMENT))
```

springstan commented:
Please take a look at the CI errors as well ✌

borpin commented:
@springstan - thoughts? I'll fix the documentation once we are settled on a solution. 😁

borpin committed:
Removed new config item

Removed the configuration item. The integration attempts to get the unit from the API.

If this fails *or* the unit key of the API is blank, either the specified unit, or the default will be used.

If approved, documentation will be updated.

borpin commented:
Third time lucky 😀. I'm learning how you do things!

Config item removed. Unit will be assigned by
1. API
2. Specified unit (if no unit key or value if blank)
3. Default Unit (if no unit specified).

I'll check the CI bits once they are complete.

I'll update the docs if this is approved.

Probably needs to be a breaking change.

springstan commented on homeassistant/components/emoncms/sensor.py:
```diff
@@ -107,6 +106,10 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
         if sensor_names is not None:
             name = sensor_names.get(int(elem["id"]), None)
 
+        unit_of_measurement = elem.get("unit", "")
```
`dict.get()` already returns None so we should work with that instead of specifying a default value as an empty string. Moreover, we should use a new variable.
```suggestion
        unit = elem.get("unit")
```

springstan commented on homeassistant/components/emoncms/sensor.py:
```diff
@@ -107,6 +106,10 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
             name = sensor_names.get(int(elem["id"]), None)
 
+        unit_of_measurement = elem.get("unit", "")
+        if unit_of_measurement == "":
```
Here we could use the `empty string` and `None` as a falsy value:
```suggestion
        if not unit:
```

springstan commented on homeassistant/components/emoncms/sensor.py:
```diff
@@ -107,6 +106,10 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
 
+        unit_of_measurement = elem.get("unit", "")
+        if unit_of_measurement == "":
+            unit_of_measurement = config.get(CONF_UNIT_OF_MEASUREMENT)
```
Since it is already set in line 77, we do not need to fetch it again. Instead we only override the fetched value if the api has returned something:
```suggestion
            unit_of_measurement = unit
```

borpin replied:
No, it needs to do it again as the variable `unit_of_measurment` at this point is the last used value not the default unit.

springstan replied:
Yeah, you are right 👍 but then we need to change it a bit to:
```python
if unit:
  unit_of_measurement = unit
else:
  unit_of_measurement = config.get(CONF_UNIT_OF_MEASUREMENT)
```

borpin replied:
Ok if you prefer that to just setting `unit_of_measurment` twice.

borpin replied:
Probably easiest if you change and I'll accept!

springstan reviewed (commented):
Changed the logic a bit to remove duplicate code and make it easier to read.

borpin committed:
Update homeassistant/components/emoncms/sensor.py

Co-Authored-By: springstan <46536646+springstan@users.noreply.github.com>

borpin committed:
Update homeassistant/components/emoncms/sensor.py

Co-Authored-By: springstan <46536646+springstan@users.noreply.github.com>

springstan commented on homeassistant/components/emoncms/sensor.py:
```diff
@@ -107,6 +106,10 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+        unit = elem.get("unit")
+        if unit:
+            unit_of_measurement = config.get(CONF_UNIT_OF_MEASUREMENT)
+
```
```suggestion
else:
    unit_of_measurement = config.get(CONF_UNIT_OF_MEASUREMENT)
```

springstan committed:
Apply suggestions from code review

springstan commented on homeassistant/components/emoncms/sensor.py:
```diff
@@ -107,6 +106,12 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+        unit = elem.get("unit")
+        if unit:
+            unit_of_measurement = unit
+else:
```
```suggestion
        else:
```

springstan commented on homeassistant/components/emoncms/sensor.py:
```diff
@@ -107,6 +106,12 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+        if unit:
+            unit_of_measurement = unit
+else:
+    unit_of_measurement = config.get(CONF_UNIT_OF_MEASUREMENT)
```
```suggestion
            unit_of_measurement = config.get(CONF_UNIT_OF_MEASUREMENT)
```

springstan committed:
Apply suggestions from code review v2

springstan commented on homeassistant/components/emoncms/sensor.py:
```diff
@@ -107,6 +106,12 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+        if unit:
+            unit_of_measurement = unit
+        else:
+            unit_of_measurement = config.get(CONF_UNIT_OF_MEASUREMENT)
```
We could even make this line easier by changing the variable in line 77 from `unit_of_measurement` to `config_unit` and then using it here. This would make things easier since we do not need to access the config multiple times. Unfortunately, I cannot change line 77 from github so you will have to rename it 😄
```suggestion
            unit_of_measurement = config_unit
```

borpin replied:
OK done (missed your last comment!).

borpin committed:
Update homeassistant/components/emoncms/sensor.py

Co-Authored-By: springstan <46536646+springstan@users.noreply.github.com>

borpin committed:
Update sensor.py

Update `config_unit`

springstan reviewed (approved):
LGTM 👌 thanks for implementing my suggested changes!
